src/google_publisher.rs

```diff
@@ -79,144 +159,483 @@ impl<C> GooglePubSubPublisher<C> {
-                const AUTH_SCOPES: [&str; 1] = ["https://www.googleapis.com/auth/pubsub"];
-                let token = authenticator
-                    .token(&AUTH_SCOPES)
-                    .await
-                    .map_err(GooglePubsubError::GetAuthToken)?;
-                let data = serde_json::to_vec(&PubsubPublishRequestSchema {
-                    messages: &messages,
-                })
-                .map_err(GooglePubsubError::SerializePublishRequest)?;
-                let request = http::Request::post(uri)
-                    .header(
-                        http::header::AUTHORIZATION,
-                        format!("Bearer {}", token.as_str()),
-                    )
-                    .header(http::header::ACCEPT, JSON_METATYPE)
-                    .header(http::header::CONTENT_TYPE, JSON_METATYPE)
-                    .body(hyper::Body::from(data))
-                    .map_err(GooglePubsubError::ConstructRequest)?;
-                let response = client
-                    .request(request)
-                    .map_err(GooglePubsubError::PostMessages)
-                    .await?;
-                let (parts, body) = response.into_parts();
-                let body_data = body
-                    .map_ok(|v| v.to_vec())
-                    .try_concat()
-                    .map_err(GooglePubsubError::ResponseBodyReceive)
-                    .await?;
-                if !parts.status.is_success() {
-                    let src = serde_json::from_slice(&body_data)
-                        .ok()
-                        .map(|v: PubsubPublishFailResponseSchema| v.error.message.into());
-                    return Err(GooglePubsubError::ResponseStatus(src, parts.status));
-                }
-                let rsp: PubsubPublishResponseSchema =
-                    serde_json::from_slice(&body_data).map_err(GooglePubsubError::ResponseParse)?;
-                Ok(rsp)
+            google_cloud_project, topic
+        ))
+        .map_err(Arc::new);
+        let uri = match uri {
+            Ok(uri) => uri,
+            Err(e) => {
+                let c = messages.count();
+                return GooglePubSubPublishStream(Box::pin(futures::stream::iter(
+                    std::iter::repeat(Err(GooglePubSubError::ConstructRequestUri(e))).take(c),
+                )));
             }
-            .await;
-            match result {
-                Ok(PubsubPublishResponseSchema { message_ids }) => {
-                    PublisherResult::Success(message_ids)
-                }
-                Err(e) => PublisherResult::OneError(e.into(), messages),
-            }
-        }))
+        };
+        let stream = GoogleMessageSegmenter::new(messages)
+            .map(|batch| {
+                publish_single_body(batch, uri.clone(), authenticator.clone(), client.clone())
+            })
+            .collect::<futures::stream::FuturesOrdered<_>>()
```

Comment:
> Is that the desired behavior, compared to sending requests serially?

Yeah. The only requirement for this code is that we retain the ordering of responses corresponding to requests made so that we can reestablish the correspondence of messages to the IDs.

> are there any concerns with bandwidth or message limits?

Nothing that's documented. Ultimately, if the uplink is unable to support the message stream there's no helping it either way, the user will get backed up either way. Running the requests concurrently at the very least utilizes most of that bandwidth.